torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -439,7 +439,28 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
           if (futureResponse->completed()) {
             --serverActiveCalls_;
             if (!futureResponse->hasError()) {
-              send(work.from_, std::move(*futureResponse).moveValue());
+              // Hack: Check for device type here since throwing an exception in
```

Comment:
Would I be correct if I assume we wouldn't reach here at all in normal cases as we have already errored out in `wireSerialize`?

Reply:
No we would reach here in normal cases if the response has a GPU tensor.

Reply:
Oh I see, it is for response. In that case shall we use the same check in `wireSerialize` and convert the error into an error message in `handleSend` for the response? So that we don't need to have the same logic duplicated in two places. 

Just to check IIUC, we do it here because this is on a separate thread so that we don't have the opportunity to directly throw the error, and hence explicitly check it and turn it into an error message?

Reply:
Good point, let me try to merge it.